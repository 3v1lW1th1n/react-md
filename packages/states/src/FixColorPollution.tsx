import React, {
  isValidElement,
  Fragment,
  FunctionComponent,
  ReactNode,
} from "react";
import useStatesContext from "./useStatesContext";

export interface FixColorPollutionProps {
  enabled?: boolean;
  children: ReactNode;
}

/**
 * This component is used to alongside the `ColorPollutionContext` and the
 * `rmd-states-prevent-color-pollution` scss variable. This will modify
 * the children of your "interactiable" element so that their text
 * color will not have the opacity applied to them as well.
 */
const FixColorPollution: FunctionComponent<FixColorPollutionProps> = ({
  enabled,
  children,
}) => {
  const isFixEnabled =
    typeof enabled === "boolean"
      ? enabled
      : useStatesContext().preventColorPollution;

  if (!isFixEnabled) {
    if (isValidElement(children)) {
      return children;
    }

    return <Fragment>{children}</Fragment>;
  }

  return (
    <Fragment>
      <span aria-hidden="true" className="rmd-states__content-mask">
        {children}
      </span>
      <span className="rmd-states__content-overlay">{children}</span>
    </Fragment>
  );
};

if (process.env.NODE_ENV !== "production") {
  let PropTypes;
  try {
    PropTypes = require("prop-types");
  } catch (e) {}

  if (PropTypes) {
    FixColorPollution.propTypes = {
      enabled: PropTypes.bool,
      children: PropTypes.node.isRequired,
    };
  }
}

export default FixColorPollution;
