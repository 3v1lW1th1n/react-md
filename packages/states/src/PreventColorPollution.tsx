import React, { createContext, FunctionComponent, ReactNode } from "react";

/**
 * This is a global context that will update all interactable elements
 * that are using the `FixColorPollution` element to render their children
 * so that the interaction states do not modify the text color. This is
 * disabled by default since it might seem confusing to modify the dom
 * structure so much to fix a weird issue out of the box. It is recommended
 * to enable it though.
 */
export const ColorPollutionContext = createContext(false);

export interface IPreventColorPollutionProps {
  enabled?: boolean;
  children: ReactNode;
}

export interface IPreventColorPollutionDefaultProps {
  enabled: boolean;
}

type PreventColorPollutionWithDefaultProps = IPreventColorPollutionProps &
  IPreventColorPollutionDefaultProps;

/**
 * This is an extremely simple component that will update the `ColorPollutionContext`
 * to have a value of `true` isntead of false.
 */
const PreventColorPollution: FunctionComponent<
  IPreventColorPollutionProps
> = props => {
  const { enabled, children } = props as PreventColorPollutionWithDefaultProps;
  return (
    <ColorPollutionContext.Provider value={enabled}>
      {children}
    </ColorPollutionContext.Provider>
  );
};

const defaultProps: IPreventColorPollutionDefaultProps = {
  enabled: true,
};

PreventColorPollution.defaultProps = defaultProps;

if (process.env.NODE_ENV !== "production") {
  let PropTypes;
  try {
    PropTypes = require("prop-types");
  } catch (e) {}

  if (PropTypes) {
    PreventColorPollution.propTypes = {
      enabled: PropTypes.bool,
      children: PropTypes.node.isRequired,
    };
  }
}

export default PreventColorPollution;
