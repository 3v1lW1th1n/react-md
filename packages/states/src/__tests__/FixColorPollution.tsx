import React from "react";
import { mount } from "enzyme";

import FixColorPollution from "../FixColorPollution";
import PreventColorPollution from "../PreventColorPollution";

describe("FixColorPollution", () => {
  it("should render the children as normal when the color pollution context is diabled", () => {
    expect(
      mount(
        <button type="button">
          <FixColorPollution>
            <span>Hello</span>
          </FixColorPollution>
        </button>
      ).render()
    ).toMatchSnapshot();
    expect(
      mount(
        <button type="button">
          <FixColorPollution enabled={false}>
            <span>Hello</span>
          </FixColorPollution>
        </button>
      ).render()
    ).toMatchSnapshot();
  });

  it("should render correctly when the fix is enabled", () => {
    // have to wrap with an additional element since enzyme can't render fragments yet
    // it also won't really be a real problem since these color fixes should always
    // be children of an element
    expect(
      mount(
        <PreventColorPollution>
          <button type="button">
            <FixColorPollution>
              <span>Hello</span>
            </FixColorPollution>
          </button>
        </PreventColorPollution>
      ).render()
    ).toMatchSnapshot();
    expect(
      mount(
        <button type="button">
          <FixColorPollution enabled>
            <span>Hello</span>
          </FixColorPollution>
        </button>
      ).render()
    ).toMatchSnapshot();
  });
});
