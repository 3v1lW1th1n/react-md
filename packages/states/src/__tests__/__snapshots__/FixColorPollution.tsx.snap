// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FixColorPollution should render correctly when the fix is enabled 1`] = `
<button
  type="button"
>
  <span
    aria-hidden="true"
    class="rmd-states__content-mask"
  >
    <span>
      Hello
    </span>
  </span>
  <span
    class="rmd-states__content-overlay"
  >
    <span>
      Hello
    </span>
  </span>
</button>
`;

exports[`FixColorPollution should render correctly when the fix is enabled 2`] = `
<button
  type="button"
>
  <span
    aria-hidden="true"
    class="rmd-states__content-mask"
  >
    <span>
      Hello
    </span>
  </span>
  <span
    class="rmd-states__content-overlay"
  >
    <span>
      Hello
    </span>
  </span>
</button>
`;

exports[`FixColorPollution should render the children as normal when the color pollution context is diabled 1`] = `
<button
  type="button"
>
  <span>
    Hello
  </span>
</button>
`;

exports[`FixColorPollution should render the children as normal when the color pollution context is diabled 2`] = `
<button
  type="button"
>
  <span>
    Hello
  </span>
</button>
`;
